Bench RaptorQ

carllin opened the pull request:
#### Problem
Reed-solomon erasure doesn't scale well over large numbers of symbols

#### Summary of Changes
Bench comparing to RaptorQ. Compute for raptorQ should scale linearly with number of symbols, Reed-solomon scales quadratically

Results on my laptop: 2.6 GHz Core i7 (I7-8850H)

```
Batch of 32 shreds, with 32:32 erasure

test bench_shredder_coding           ... bench:     530,523 ns/iter (+/- 133,574)
test bench_shredder_coding_raptorq   ... bench:     207,934 ns/iter (+/- 77,269)

test bench_shredder_decoding         ... bench:     821,487 ns/iter (+/- 38,513)
test bench_shredder_decoding_raptorq ... bench:     211,483 ns/iter (+/- 6,725)
```

```
Batch of 32 shreds, with 32:64 erasure
test bench_shredder_coding_raptorq   ... bench:     217,074 ns/iter (+/- 16,993)
test bench_shredder_decoding_raptorq ... bench:     209,287 ns/iter (+/- 5,525)
```

```
Batch of 32 shreds, with 32:96 erasure
test bench_shredder_coding_raptorq   ... bench:     225,000 ns/iter (+/- 8,226)
test bench_shredder_decoding_raptorq ... bench:     212,213 ns/iter (+/- 13,668)
```

```
Batch of 1000 shreds, with 1000:1000 erasure
test bench_shredder_coding_raptorq   ... bench:   6,313,892 ns/iter (+/- 3,814,892)
test bench_shredder_decoding_raptorq ... bench:   6,656,319 ns/iter (+/- 1,135,365)
```

Results suggest that:
1)  RaptorQ coding for our current use case `32:32` is about 2.5x faster in encoding, about 4x faster in decoding.
2) It's very cheap to generate higher coding ratios `1:2` `1:3` that are currently untenable in our current erasure scheme (don't know how useful this is other than padding small FEC blocks that only have 1 or 2 shreds which are susceptible to packet drops).
3) RaptorQ efficiently generates larger FEC blocks, for instance if we wanted to have `100:100`, or `1000:1000` at higher TPS. 

Fixes #

ryoqun commented:
I'm not quite familiar with Reed-solomon and RaptorQ. But how these performance wins are related to our past discussion there?: https://github.com/solana-labs/solana/pull/4352.

carllin commented:
@ryoqun, some of that that discussion looks quite outdated as we've restructured a lot of the logic surrounding coding shred generation and storage in blockstore. From what I understand currently, the RaptorQ stuff can mostly be a drag and drop for the Reed-Solomon stuff we currently have implemented. 

The primary difference is that in Reed-Solomon if you encode a source of `k` original symbols, then receiving any `k` out of `n` of the symbols is 100% guaranteed to recover.

With RaptorQ, receiving any `k` out of `n` of the symbols will probabilistically allow recovery, with each additional symbol decreasing probability of failure by a factor of `256`. From the description in the README: https://github.com/cberner/raptorq

```
Recovery properties: Reconstruction probability after receiving K + h packets = 1 - 1/256^(h + 1). Where K is the number of packets in the original message, and h is the number of additional packets received. 
```

Addressing some points from #4352.:
1) We don't currently repair erasure shreds, so this would not be affected by a swap 
2) Signing the raptor shreds may have a perf impact in broadcast if we decide to go with a greater source to coding than we have today (currently ratio is `1:1`). Since that discussion though we've offloaded signing to separate threads for better pipelining, so it may be a nonissue.
